Fixed clock.fakeTimer.Stop and Reset

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Fixed the way clock.fakeTimer.Stop() identified the fakeClockWaiter to remove.  The previous code never removed any fakeClockWaiter, so Stop() had no effect.

Based the return value on whether the timer's fakeClockWaiter remains among those of the fake clock.  The previous code always returned `true`, due to examining a copy of the original state of the fakeClockWaiter.

Also fixed clock.fakeTimer.Reset(), it suffered from the same fundamental misconception.

Added a test func for FakeClock timers.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #78807

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

MikeSpreitzer committed:
Fixed clock.fakeTimer.Stop and Reset

Fixed the way Stop identified the fakeClockWaiter to remove, and the
logic for the bool to return.

Fixed Reset to side-effect the waiter in the clock's slice rather than
the original copy.

Added a test func for FakeClock timers.

Fixed Reset too

Improved comments in clock.go to address reviewer comments

Hopefully the updated comments make clear the answers to recent
reviewer comments.

MikeSpreitzer commented:
@kubernetes/sig-api-machinery-bugs

ncdc commented on staging/src/k8s.io/apimachinery/pkg/util/clock/clock_test.go:
```diff
@@ -117,6 +117,67 @@ func TestFakeAfter(t *testing.T) {
+		t.Errorf("unexpected channel read")
+	default:
+	}
+	if oneSec.Stop() {
```
Shouldn't this also check for `false` because `oneSec` has already fired?

ncdc replied:
EDIT: nevermind!

ncdc commented on staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go:
```diff
@@ -291,18 +291,19 @@ func (f *fakeTimer) C() <-chan time.Time {
 	f.fakeClock.lock.Lock()
 	defer f.fakeClock.lock.Unlock()
-
+	fired := true
```
Would this be clearer and/or cleaner as:

```go
fired := f.waiter.fired
newWaiters := make([]fakeClockWaiter, 0, len(f.fakeClock.waiters))
for _, w := range f.fakeClock.waiters {
  if w.destChan != f.waiter.destChan {
    newWaiters = append(newWaiters, w)
  }
}

ncdc commented:
@MikeSpreitzer thanks for putting together this fix!

/lgtm

fedebongio commented:
/assign @lavalamp

fedebongio commented:
/assign @logicalhan

MikeSpreitzer commented:
/assign @liggitt

logicalhan commented on staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go:
```diff
@@ -291,18 +291,19 @@ func (f *fakeTimer) C() <-chan time.Time {
 	f.fakeClock.lock.Lock()
 	defer f.fakeClock.lock.Unlock()
-
+	fired := true
```
I prefer `fired := false` here and setting `fired = true` on line 298 if you find the waiter channel. That would simplify the return statement (no negation necessary).

MikeSpreitzer replied:
I am not sure I understand.  `Stop()` is defined to return `true` if the timer has _not_ fired.  If we want `Stop()` to return the value of a variable without negation then that variable should be named something like "notFired" --- and we still have a negation but it is in the variable name rather than the logic.  In my opinion that makes the logic more difficult to read.  Also, changing line 298 to ignore `w.fired` strikes me as a brittle coding even though it looks correct at the moment.

logicalhan replied:
It seems that either the destChan is there, in which case the fired attribute is false, or if it has fired then it shouldn't actually be present so our equality check won't occur. Why should we have a layer of indirection here?

MikeSpreitzer replied:
No, actually, `w.fired` might be true so we positively need line 298 to read it.

logicalhan reviewed (commented):
/lgtm

MikeSpreitzer commented:
On further consideration and discussion with @logicalhan , I realized that (a) the fake timer's Reset() was broken too and (b) we do not need the `fired` field.  I fixed those things, and rebased, and thus force-pushed.

logicalhan reviewed (approved):
/lgtm

Thanks for fixing these!

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go:
```diff
@@ -291,32 +288,35 @@ func (f *fakeTimer) C() <-chan time.Time {
 	f.fakeClock.waiters = newWaiters
 
-	return !f.waiter.fired
+	return stopped
```
Stop() is supposed to return false if it was already stopped, or if the timer had already fired. I'm having a hard time understanding how the channel equality comparison is accomplishing the latter.

MikeSpreitzer replied:
When the timer fires or is stopped, it is removed from the clock's list.  The channel equality test is how the timer's entry is identified in the list.  The fakeClockWaiter does not retain a constant memory address, so a pointer to a fakeClockWaiter can not be used to identify the one belonging to the timer.

MikeSpreitzer replied:
See if the new revision includes comments that answer this question.

logicalhan replied:
@liggitt if you are confused now, it was even more confusing before @MikeSpreitzer made these improvements. There was a fired attribute on the waiter struct which only ever had one value, which was that it hadn't fired. This seems markedly better to me.

MikeSpreitzer commented:
I pushed a new revision that should make clear the answer to @liggitt's comment, and also help with some things raised in the review by @logicalhan (whose LGTM has thus been shorn off; can you PTAL?).  I noticed that the comments on fakeTimer::Stop and fakeTimer::Reset were not accurate, so I updated them to be accurate and also to reference the source of their design, time.Timer.  I also added a couple of comments into fakeTimer::Stop to explain the implementation.  I eschewed adding implementation comments into fakeTimer::Reset, hoping that the comments in fakeTimer::Stop would be enough.

ncdc commented on staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go:
```diff
@@ -287,36 +284,48 @@ func (f *fakeTimer) C() <-chan time.Time {
 
-// Stop stops the timer and returns true if the timer has not yet fired, or false otherwise.
+// Stop conditionally stops the timer.  If the timer has neither fired
+// nor been stopped then this call stops the timer and returns true,.
```
```suggestion
// nor been stopped then this call stops the timer and returns true,
```

ncdc commented on staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go:
```diff
@@ -287,36 +284,48 @@ func (f *fakeTimer) C() <-chan time.Time {
-		if w != &f.waiter {
-			newWaiters = append(newWaiters, *w)
+	for _, w := range f.fakeClock.waiters {
+		// Identify the timer's fakeClockWaiter by the identify of the
```
```suggestion
		// Identify the timer's fakeClockWaiter by the identity of the
```

logicalhan reviewed (approved):
/lgtm

MikeSpreitzer commented:
Oh dang.  Another typo fix and squash coming.

logicalhan reviewed (commented):
/lgtm

logicalhan commented:
/priority important-soon

MikeSpreitzer commented:
@ncdc , @logicalhan : all typos fixed now, sorry, need your LGTM again.  @liggitt : the comments as well as the implementation should be good now.

MikeSpreitzer commented:
No, wait, still git confusion on my end.

MikeSpreitzer commented:
Nope! _still_ git confusion

MikeSpreitzer commented:
OK, @ncdc , @logicalhan : I have _finally_ won my git battle and it is ready for final review.

logicalhan reviewed (commented):
/lgtm

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
/test pull-kubernetes-bazel-test

MikeSpreitzer commented:
The pull-kubernetes-kubemark-e2e-gce-big job failure must also be a flake.  This PR changes only the behavior of unit tests.

liggitt commented:
/approve
/retest
